### DISCUSSION TITLE
Nested and globally available user script

### DISCUSSION THREAD

**Discussion by YlanAllouche** [SilentVoid13/Templater#1511] [Templates Showcase]
Hey, 
I don't know how obvious this is to everyone but you can call user modules from other user module once you pass the `tp` object as a context.

This allows breaking down functions into reusable atoms.

Even better, because Templater registers the CommonJs module in the script folder into the API, it means those modules are also globally available from anywhere else making Templater probably the best way to expose the Obsidan API in order to build a personal repository of functions without necessarily relying on a custom plugin.

For example: the content of `scripts/myJSfile.js`:

```js
function functionName () {
    console.log('I execute from that first file')
}
module.exports = functionName;

```
then `scripts/second.js` with this content.

```js
function functionName (tp, secondArgument) {
    tp.user.myJSfile() 
    console.log(secondArgument)
}
module.exports = functionName;
```

then in template

```js
<% tp.user.second(tp, "some parameter") %>
```

But this also works anywhere in obsidian:
```js
let tp = app.plugins.plugins["templater-obsidian"].templater
tp.user.second(tp, "some parameter")
```

I have a write up [here](https://www.ylanallouche.xyz/posts/global-register-func-obsidian/) to explore that and more (it's also great for remote execution).

**Comment by ca11ado** [comment]
Also, I needed the ability to share some state between functions, and I managed to solve this problem with the following approach:

utils.js
```
const SHARED_CONTEXT = ['project', 'feature'];

function my_first_function() {
  return SHARED_CONTEXT;
}

function my_second_function() {
  return SHARED_CONTEXT;
}

function fn() {
  return {
    my_first_function,
    my_second_function,
  };
}

module.exports = fn;
```

After that you can call your functions as this:
```
<%
  tp.user.fn().my_first_function();
  tp.user.fn().my_second_function();
%>
```

**Reply by NicholasGGW** [reply]
thx,this helps a lot!My test scripts is like
math.js
```
function add(a, b) {
    return a + b;
}

function subtract(a, b) {
  return a - b;
}

function multiply(a, b) {
  return a * b;
}

function fn() {
    return {
        add,
        subtract,
        multiply
    };
}

module.exports = fn;
```

and I select the script folder location in Templater settings
use
```
<%* tp.user.math().subtract(2,3) %>
<%* 
	tR += tp.user.math().multiply(2,3)
%>
```
to call mutiply functions in one js file!
